Source/DafnyDriver/Commands/DafnyCli.cs

```diff
@@ -0,0 +1,514 @@
+    Contract.Requires(cce.NonNullElements(args));
+    return Execute(outputWriter, errorWriter, inputReader, args, async parseArgumentResult => {
+
+      switch (parseArgumentResult) {
+        case ParsedOptions success:
+          var options = success.DafnyOptions;
+          return await CompilerDriver.RunCompiler(options);
+        case ExitImmediately failure:
+          return (int)failure.ExitValue;
+        default: throw new Exception("unreachable");
+      }
+    });
+  }
+
+  private static void AddCommand(Command command) {
+    RootCommand.AddCommand(command);
+  }
+
+  static DafnyCli() {
+    AddCommand(ResolveCommand.Create());
+    AddCommand(VerifyCommand.Create());
+    AddCommand(BuildCommand.Create());
+    AddCommand(RunCommand.Create());
+    AddCommand(TranslateCommand.Create());
+    AddCommand(FormatCommand.Create());
+    AddCommand(DocCommand.Create());
+    AddCommand(MeasureComplexityCommand.Create());
+    AddCommand(ServerCommand.Instance);
+    AddCommand(TestCommand.Create());
+    AddCommand(GenerateTestsCommand.Create());
+    AddCommand(DeadCodeCommand.Create());
+    AddCommand(AuditCommand.Create());
+    AddCommand(CoverageReportCommand.Create());
+
+    // Check that the .doo file format is aware of all options,
+    // and therefore which have to be saved to safely support separate verification/compilation.
+    DooFile.CheckOptions(AllOptions);
+
+    // This SHOULD find the same method but returns null for some reason:
+    // typeof(ParseResult).GetMethod("GetValueForOption", 1, new[] { typeof(Option<>) });
+    foreach (var method in typeof(ParseResult).GetMethods()) {
+      if (method.Name == "GetValueForOption" && method.GetGenericArguments().Length == 1) {
+        GetValueForOptionMethod = method;
+      }
+    }
+    Debug.Assert(GetValueForOptionMethod != null);
+
+    var builder = new CommandLineBuilder(RootCommand).UseDefaults();
+    builder = AddDeveloperHelp(RootCommand, builder);
+    Parser = builder.Build();
+  }
+
+  class WritersConsole : IConsole {
+    public TextReader InputWriter { get; }
+    public TextWriter ErrWriter { get; }
+    public TextWriter OutWriter { get; }
+
+    public WritersConsole(TextReader inputWriter, TextWriter outWriter, TextWriter errWriter) {
+      InputWriter = inputWriter;
+      this.ErrWriter = errWriter;
+      this.OutWriter = outWriter;
+    }
+
+    public IStandardStreamWriter Out => StandardStreamWriter.Create(OutWriter ?? TextWriter.Null);
+
+    public bool IsOutputRedirected => OutWriter != null;
+    public IStandardStreamWriter Error => StandardStreamWriter.Create(ErrWriter ?? TextWriter.Null);
+    public bool IsErrorRedirected => ErrWriter != null;
+    public bool IsInputRedirected => false;
+  }
+
+  public delegate Task<int> ContinueWithOptions(DafnyOptions dafnyOptions, InvocationContext context);
+  public static void SetHandlerUsingDafnyOptionsContinuation(Command command, ContinueWithOptions continuation) {
```

Comment:
This doesn't feel like it justifies using "continuation" as in CPS. Perhaps just `TaskWithOptions` and `SetCommandHandlerImplementation`?

Reply:
From wiki:

> A function written in continuation-passing style takes an extra argument: an explicit "continuation"; i.e., a function of one argument. When the CPS function has computed its result value, it "returns" it by calling the continuation function with this value as the argument.

The function `SetHandlerUsingDafnyOptionsContinuation` takes an extra argument, the function `continuation`. When the handler has computed the DafnyOptions value, it "returns" it by calling the function `continuation` with this value as an argument.

To rename `ContinueWithOptions` to `TaskWithOptions` would be incorrect because `ContinueWithOptions` is not a task: it's a function that returns a task. `SetCommandHandlerImplementation` sounds like it does the same as `command.SetHandler`, which already exists.

Instead of `continuation` I can also use the term callback.

Reply:
Okay I see it now, `SetHandlerUsingDafnyOptionsContinuation` itself isn't really in CPS style, but the handler beting set is.

> Instead of continuation I can also use the term callback.

That might be slightly better but I'm okay with this as is.